### PR TITLE
Add autosave feature

### PR DESCRIPTION
**Pull request by superenginegit**
This PR is for the feature request of issue #1926.

**What I implemented:**
- Settings for autosaves: Autosaving can be toggled on or off and the time interval between two autosaves can be set.
- Creation of autosaves: Autosaving is performed regularly, but a new autosave will only be created if there were changes since the last autosave. Autosaves will be made for files that have been saved, for those they will be created in the same location as the original file. Autosave for files that have never be saved will be created in the users home directory.
- Autosaves that have been created for unsaved files will be checked for on start of logisim. And they can then be saved.
- Autosaves for saved files will be checked when that file is reopened. It can then be restored or deleted.

### CODE REVIEW

File: src/main/java/com/cburch/logisim/file/Loader.java [hunk 1]
```diff
@@ -142,6 +149,28 @@ private static File determineBackupName(File base) {
     return null;
   }
 
+  // Determine the autosave file name for any .circ file as .<basename>.circ.autosave
+  private static File determineAutosaveName(File base) {
+    if (base == null) {
+      String timestamp =
+          DateTimeFormatter.ofPattern("yyyy-MM-dd_HH:mm:ss").format(LocalDateTime.now());
```
Comment (maehne) [inline src/main/java/com/cburch/logisim/file/Loader.java:156]: `:` in file names are not universally allowed, e.g., Windows doesn't allow them. Personally, I would opt to use `yyyyMMddHHmmss`, which also enables chronological sorting.

File: src/main/java/com/cburch/logisim/file/Loader.java [hunk 2]
```diff
@@ -419,9 +449,50 @@ public boolean save(LogisimFile file, File dest) {
...
+    return true;
+  }
+
+  /**
+   * Method to perform autosaves. Essentially does the same as save()
+   * but without any failsafes, if saving fails it simply fails.
+   *
+   * @param file The file that should be autosaved
+   *
+   * @return True if writing was successfull, else false;
```
Comment (maehne) [inline src/main/java/com/cburch/logisim/file/Loader.java:464]: Typo: "successfull" -> "successful"

File: src/main/java/com/cburch/logisim/file/LogisimFile.java [hunk 3]
```diff
@@ -163,10 +214,32 @@ private static String getFirstLine(BufferedInputStream in) throws IOException {
...
+      final var res = loader.showOptions(S.get("contentHandleAutosave", file.getName()),
+          S.get("titleHandleAutosave"), new String[] {S.get("loadOption"), S.get("discardOption")},
+          0);
+
+      if (res == JOptionPane.CLOSED_OPTION) { // If the prompt was closed do nothing and fail
+        return null;
+      } else if (res == 0) { // If load is selected select the autosave to be loaded
+        loadFile = autosave.get(); // Set load file to the autosave path
+        loader.setAutosavePath(autosave.get());
+        autosaveLoading = true; // also set this to true to remember an autosave was loaded
```
Comment (maehne) [inline src/main/java/com/cburch/logisim/file/LogisimFile.java:232]: "also" -> "Also" to be consistent with casing in other comments?

File: src/main/resources/resources/logisim/strings/file/file.properties [hunk 4]
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = Library includes items currently in the toolbar.
 unloadUsedError = Circuit '%s' uses components from the library.
 xmlConversionError = Internal error while creating XML
 xmlFormatError = XML formatting error: %s
+autosaveError = Failed to create autosave for file: '%s'. Will not try again for this File until after a restart.
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/file/file.properties:47]: "File" -> "file"

**Review by maehne — CHANGES_REQUESTED** [review]
Thanks for implementing this useful feature! It looks mostly good for me, but as this modifies some quite sensitive parts of the code, it should be at least also reviewed by @BFH-ktt1. It would be nice if you could add the newly introduced localisation strings to alls supported languages as placeholders at least. This facilitates additional translations.

**Comment by superenginegit** [comment]
Thanks for the review, I have now added translation placeholders to the languages and fixed the other things you mentioned as well.

File: src/main/resources/resources/logisim/strings/file/file_fr.properties [hunk 5]
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = La librairie inclut des éléments de la barre d'outils.
 unloadUsedError = Le circuit '%s' utilise des composants de la librarie.
 xmlConversionError = Erreur interne pendant la création du fichier XML
 xmlFormatError = Erreur de formatage XML : %s
+# ==> autosaveError = Failed to create autosave for file: '%s'. Will not try again for this file until after a restart.
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/file/file_fr.properties:47]: "Échec de la création d'une sauvegarde automatique pour le fichier : '%s'. La tentative ne sera pas renouvelée pour ce fichier avant un redémarrage."

File: src/main/resources/resources/logisim/strings/file/file_fr.properties [hunk 6]
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = La librairie inclut des éléments de la barre d'outils.
...
+# ==> titleHandleAutosave = Autosave found
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/file/file_fr.properties:48]: "Sauvegarde automatique trouvée"

File: src/main/resources/resources/logisim/strings/file/file_fr.properties [hunk 7]
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = La librairie inclut des éléments de la barre d'outils.
...
+# ==> contentHandleAutosave = Autosave for file '%s' found. What should be done?
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/file/file_fr.properties:49]: "Sauvegarde automatique pour le fichier '%s' trouvée. Que faut-il faire ?"

File: src/main/resources/resources/logisim/strings/file/file_fr.properties [hunk 8]
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = La librairie inclut des éléments de la barre d'outils.
...
+# ==> loadOption = Load
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/file/file_fr.properties:50]: "Charger"

File: src/main/resources/resources/logisim/strings/file/file_fr.properties [hunk 9]
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = La librairie inclut des éléments de la barre d'outils.
...
+# ==> discardOption = discard
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/file/file_fr.properties:51]: "Écarter"

File: src/main/resources/resources/logisim/strings/file/file_nl.properties
Comment (maehne) [inline src/main/resources/resources/logisim/strings/file/file_nl.properties:1]: @BFH-ktt1: Maybe you could provide translations for these strings?

File: src/main/resources/resources/logisim/strings/gui/gui_fr.properties [hunk 10]
```diff
@@ -600,6 +600,13 @@ templateSelectButton = Sélectionner...
 templateTitle = Modèle
 templateRemoveLibsOnSave = Supprimer les bibliothèques inutilisées lors de l'enregistrement
 #
+# prefs/AutosaveOptions.java
+#
+# ==> autosaveEnabled = Enable autosaving (Takes effect after a restart).
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/gui/gui_fr.properties:605]: "Activer la sauvegarde automatique (prend effet après un redémarrage)."

File: src/main/resources/resources/logisim/strings/gui/gui_fr.properties [hunk 11]
```diff
@@ -600,6 +600,13 @@ templateSelectButton = Sélectionner...
...
+# ==> autosaveInterval = Autosaving interval in seconds:
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/gui/gui_fr.properties:606]: "Intervalle de sauvegarde automatique en secondes :"

File: src/main/resources/resources/logisim/strings/gui/gui_fr.properties [hunk 12]
```diff
@@ -600,6 +600,13 @@ templateSelectButton = Sélectionner...
...
+# ==> autosaveTitle = Autosave
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/gui/gui_fr.properties:607]: "Sauvegarde automatique"

File: src/main/resources/resources/logisim/strings/gui/gui_fr.properties [hunk 13]
```diff
@@ -600,6 +600,13 @@ templateSelectButton = Sélectionner...
...
+# ==> autosaveHelp = Select autosave settings
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/gui/gui_fr.properties:608]: "Sélectionner les paramètres de sauvegarde automatique"

File: src/main/resources/resources/logisim/strings/gui/gui_fr.properties [hunk 14]
```diff
@@ -743,6 +750,9 @@ templateCannotReadError = Pas la permission de lire le fichier modèle %s.
 templateMissingError = Le fichier modèle %s n'existe pas.
 ttyFormatError = --tty requière au moins l'une des options : halt, speed, stats, table, tty, binary, hex, csv, tabs
 ttyNeedsFileError = L'utilisation de "--tty" requière un nom de fichier sur la ligne de commande.
+# ==> saveAndLoadOption = Save and load
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/gui/gui_fr.properties:753]: "Sauvegarder et charger"

File: src/main/resources/resources/logisim/strings/gui/gui_fr.properties [hunk 15]
```diff
@@ -743,6 +750,9 @@ templateCannotReadError = Pas la permission de lire le fichier modèle %s.
...
+# ==> titleHandleAutosave = Autosave found
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/gui/gui_fr.properties:754]: "Sauvegarde automatique trouvée"

File: src/main/resources/resources/logisim/strings/gui/gui_fr.properties [hunk 16]
```diff
@@ -743,6 +750,9 @@ templateCannotReadError = Pas la permission de lire le fichier modèle %s.
...
+# ==> contentHandleAutosave = Autosave file '%s' found. What should be done?
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/gui/gui_fr.properties:755]: "Fichier de sauvegarde automatique '%s' trouvé. Que faut-il faire ?"

File: src/main/resources/resources/logisim/strings/file/file_de.properties [hunk 17]
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = Bibliothek enthält Elemente auf der Werkzeugleiste.
 unloadUsedError = Schaltung "%s" verwendet Elemente der Bibliothek.
 xmlConversionError = Interner Fehler beim Erstellen von XML
 xmlFormatError = Fehler beim XML-Format: %s
+autosaveError = Erstellen der Autospeicherung für Datei '%s' fehlgeschlagen. Wird bis zum Neustart nicht wiederholt.
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/file/file_de.properties:47]: "Automatische Sicherung für Datei konnte nicht erstellt werden: '%s'. Wird erst nach einem Neustart erneut für diese Datei versucht."

File: src/main/resources/resources/logisim/strings/file/file_de.properties [hunk 18]
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = Bibliothek enthält Elemente auf der Werkzeugleiste.
...
+titleHandleAutosave = Autospeicherung gefunden
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/file/file_de.properties:48]: "Automatische Sicherung gefunden"

File: src/main/resources/resources/logisim/strings/file/file_de.properties [hunk 19]
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = Bibliothek enthält Elemente auf der Werkzeugleiste.
...
+contentHandleAutosave = Autospeicherung für Datei '%s' gefunden. Wie vorgehen?
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/file/file_de.properties:49]: Automatische Sicherung für Datei '%s' gefunden. Wie vorgehen?

File: src/main/resources/resources/logisim/strings/gui/gui_de.properties [hunk 20]
```diff
@@ -600,6 +600,13 @@ templateSelectButton = Auswählen...
 templateTitle = Vorlage
 # ==> templateRemoveLibsOnSave =
 #
+# prefs/AutosaveOptions.java
+#
+autosaveEnabled = Aktiviere Autospeicherungen (Erfordert Neustart).
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/gui/gui_de.properties:605]: Aktiviere automatische Sicherungen (erfordert Neustart).

File: src/main/resources/resources/logisim/strings/gui/gui_de.properties [hunk 21]
```diff
@@ -600,6 +600,13 @@ templateSelectButton = Auswählen...
...
+autosaveInterval = Autospeicherungsinterval in Sekunden:
+autosaveTitle = Autospeichern
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/gui/gui_de.properties:607]: "Automatische Sicherungen"

File: src/main/resources/resources/logisim/strings/gui/gui_de.properties [hunk 22]
```diff
@@ -600,6 +600,13 @@ templateSelectButton = Auswählen...
...
+autosaveHelp = Autospeicherungseinstellungen ündern
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/gui/gui_de.properties]: "Einstellungen für die automatischen Sicherungen ändern"

File: src/main/resources/resources/logisim/strings/gui/gui_de.properties [hunk 23]
```diff
@@ -743,6 +750,9 @@ startupQuitButton = Beenden
 # ==> templateMissingError =
 # ==> ttyFormatError =
 # ==> ttyNeedsFileError =
+saveAndLoadOption = Speichern und laden
+titleHandleAutosave = Autospeicherung gefunden
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/gui/gui_de.properties:754]: "Automatische Sicherung gefunden"

File: src/main/resources/resources/logisim/strings/gui/gui_de.properties [hunk 24]
```diff
@@ -743,6 +750,9 @@ startupQuitButton = Beenden
...
+contentHandleAutosave = Autospeicherungsdatei '%s' gefunden. Wie vorgehen?
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/gui/gui_de.properties:755]: Automatische Sicherungsdatei '%s' gefunden. Wie vorgehen?

**Review by maehne — CHANGES_REQUESTED** [review]
Thanks for taking into account my review comments! I suggest some further improvements to the translations.

**Review by davidhutchens — COMMENTED** [review]
I looked through the code and ran some tests to see how this worked. It looks good to me. Thank you for this contribution. I have one small request.

Would you please add a line to CHANGES.md to note the addition of an autosave feature?

**Comment by superenginegit** [comment]
I've implemented the translation strings you've provided. And I've also added a line to the changes file noting the addition of autosaves. Thanks for your feedback!

**Review by maehne — APPROVED** [review]
From my side this PR is ready for merging. Unless @BFH-ktt1 would still like to provide Dutch localization strings.

**Pull request merged**